### PR TITLE
Add CALIOP v3 HDF4 reader

### PR DESCRIPTION
**Pull request by mitkin**
Currently when I load the dataset by `sc = Scene(sensor='caliop', start_time=time_slot)` and then `scn.load(['elevation'])` I get the `uknown dataset 'elevation'` error. So it looks like I am not getting to actually reading the data yet.

### CODE REVIEW

**Comment by mitkin** [comment]
The file I am trying to read

https://www.dropbox.com/s/0y24bfoveczh9sq/CAL_LID_L2_01kmCLay-ValStage1-V3-30.2015-07-11T20-17-22ZN.hdf?dl=1

The filename is: `CAL_LID_L2_01kmCLay-ValStage1-V3-30.2015-07-11T20-17-22ZN.hdf`

**Comment by davidh-ssec** [comment]
Try removing the quotes around the `name: 'elevation'` in the yaml file

**Comment by mitkin** [comment]
@davidh-ssec , quotes don't change anything unfortunately

File: satpy/readers/hdf4_caliopv3.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: Please add license and copyright info

File: satpy/readers/hdf4_caliopv3.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from pyhdf.SD import SD, SDC
+
+from satpy.dataset import Dataset
+from satpy.readers.file_handlers import BaseFileHandler
+
+logger = logging.getLogger(__name__)
+
+
+class HDF4BandReader(BaseFileHandler):
+
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: Documentation

File: satpy/readers/hdf4_caliopv3.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        self.get_filehandle()
+
+        self._start_time = filename_info['start_time']
+
+        logger.debug('Retrieving end time from metadata array')
+        self.get_end_time()
+
+    def get_end_time(self):
+        """ Get observation end time from file metadata
+        """
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: One line docstrings should end with a period, and have closing """ on the same line

File: satpy/readers/hdf4_caliopv3.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    def parse_metadata_string(metadata_string):
+        """Grab end time with regular expression
+        """
+        regex = r"STOP_DATE.+?VALUE\s*=\s*\"(.+?)\""
+        match = re.search(regex, metadata_string, re.DOTALL)
+        end_time_str = match.group(1)
+        return end_time_str
+
+    def get_filehandle(self):
+        if os.path.exists(self.filename):
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: Docstring missing

File: satpy/readers/hdf4_caliopv3.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            self.filehandle = SD(self.filename, SDC.READ)
+            logger.debug("Loading dataset {}".format(self.filename))
+        else:
+            raise IOError("Path {} does not exist.".format(self.filename))
+
+    def get_dataset(self, key, info):
+        """Read data from file and return the corresponding projectables.
+        """
+
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: No blank line between docstring and code

File: satpy/readers/hdf4_caliopv3.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                proj = Dataset(self.lons, id=key, **info)
+
+        else:
+            data = self.get_sds_variable(key.name)
+            proj = Dataset(data, id=key, **info)
+
+        return proj
+
+    def get_sds_variable(self, name):
+        """ Read variable from the HDF4 file
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: One line docstrings should end with a period, and have closing """ on the same line

File: satpy/readers/hdf4_caliopv3.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        sds_obj = self.filehandle.select(name)
+        data = sds_obj.get()
+        return data
+
+    def get_lonlats(self):
+        longitudes = self.get_sds_variable('Longitude')
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: Missing docstring

File: satpy/readers/hdf4_caliopv3.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+        logger.debug('Retrieving end time from metadata array')
+        self.get_end_time()
+
+    def get_end_time(self):
+        """ Get observation end time from file metadata
+        """
+        mda_dict = self.filehandle.attributes()
+        core_mda = mda_dict['coremetadata']
+        end_time_str = self.parse_metadata_string(core_mda)
```
Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: For parsing the metadata string, you could use hdf_eos.py's EOS metadata parser.

Comment (mitkin) [inline satpy/readers/hdf4_caliopv3.py]: Is the format the same or it needs to be overriden?
CALIOP has smth like:

```
GROUP                  = TEMPORALINFORMATION

OBJECT                 = PRODUCTIONDATETIME
  NUM_VAL              = 1
  VALUE                = "2015-07-19T18:29:43Z"
END_OBJECT             = PRODUCTIONDATETIME
```

Comment (mraspaud) [inline satpy/readers/hdf4_caliopv3.py]: It should be compatible

Comment (mitkin) [inline satpy/readers/hdf4_caliopv3.py]: The `read_mda` method would need decoupling from the HDFEOSFileReader, so I would leave the metadata parsing as it is here for now

**Pull request merged**